dict: avoid hash calculation when hash_size=1 (link list imp)

tshacked opened the pull request:
Currently dict_t always constructs with dict::hash_size = 1.
With this initializing the dict implemented as a link-list
and searching for a key is done by iteration using key comparison.
Therfore we can avoid the hash-calculation done each set()/get() in this case.

Fixes: #2013

Change-Id: Id93286a8036064d43142bc2b2f8d5a3be4e97fc4
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked committed:
dict: avoid hash calculation when hash_size=1 (link list imp)

Currently dict_t always constructs with dict::hash_size = 1.
With this initializing the dict implemented as a link-list
and searching for a key is done by iteration using key comparison.
Therfore we can avoid the hash-calculation done each set()/get() in this case.

Fixes: #2013

Change-Id: Id93286a8036064d43142bc2b2f8d5a3be4e97fc4
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
This change only prevents the hash-calculation when dict::has size = 1 (which is the code usage).
If acceptable by the community we can clean the structure and leave only the code relevant to linked-list

mykaul commented:
I thought we'd actually ifdef out the whole code that is not needed if the hash size is fixed at 1.

tshacked committed:
dict: avoid hash calculation when hash_size=1 (list imp)

Currently dict_t always constructs with dict::hash_size = 1.
With this initializing the dict implemented as a link-list
and searching for a key is done by iteration using key comparison.
Therfore we can avoid the hash-calculation done each set()/get() in this case.

Fix:
using new macro to delimit and avoid blocks related to hash imp

Fixes: #2013

Change-Id: I31180b434a6e9e7bbb456c7ad888c147c4ce3308
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
/run regression

mohit84 commented on libglusterfs/src/dict.c:
```diff
@@ -25,6 +25,12 @@
+ * with this usage it's implemented as a doubly-linked list
+ * and the hash calculation done per operation is not necessary.
+ * using this macro to delimit blocks related to hash imp */
+#define DICT_LIST_IMP 1
```
I think more simple way is  we can directly check the value of hash_size, if hash_size is 1 then we can skip hash calculation.

mohit84 replied:
I did not read yaniv comment to use a macro to ignore hash calculation if size is 1 so you can ignore my previous comment.
